### PR TITLE
fix payload index out of range error

### PR DESCRIPTION
**Pull request by chaosma**

### CODE REVIEW

File: consensus/consensus_v2.go [hunk 1]
```diff
@@ -330,44 +330,27 @@ func (consensus *Consensus) onPrepared(msg *msg_pb.Message) {
...
-	// 48 byte of multi-sig
-	multiSig := messagePayload[offset : offset+48]
-	offset += 48
 
-	// bitmap
-	bitmap := messagePayload[offset:]
-	//#### END Read payload data
+	aggSig, mask, err := consensus.readSignatureBitmapPayload(recvMsg.Payload, 0)
+	if err != nil {
+		utils.GetLogger().Debug("readSignatureBitmapPayload failed", "error", err)
```
Comment (LeoHChen) [inline consensus/consensus_v2.go]: use Error, instead of Debug

File: consensus/consensus_v2.go [hunk 2]
```diff
@@ -562,48 +545,33 @@ func (consensus *Consensus) onCommitted(msg *msg_pb.Message) {
...
-	// 48 byte of multi-sig
-	multiSig := messagePayload[offset : offset+48]
-	offset += 48
-
-	// bitmap
-	bitmap := messagePayload[offset:]
-	//#### END Read payload data
+	aggSig, mask, err := consensus.readSignatureBitmapPayload(recvMsg.Payload, 0)
+	if err != nil {
+		utils.GetLogger().Debug("readSignatureBitmapPayload failed", "error", err)
```
Comment (LeoHChen) [inline consensus/consensus_v2.go]: same here, use Error, not Debug.

File: consensus/view_change.go [hunk 3]
```diff
@@ -258,28 +258,20 @@ func (consensus *Consensus) onViewChange(msg *msg_pb.Message) {
...
-			err = deserializedMultiSig.Deserialize(multiSig)
+			if len(recvMsg.Payload) <= 32 {
+				utils.GetLogger().Debug("m1 recvMsg payload not enough length", "len", len(recvMsg.Payload))
+				return
+			}
+			blockHash := recvMsg.Payload[:32]
+			aggSig, mask, err := consensus.readSignatureBitmapPayload(recvMsg.Payload, 32)
 			if err != nil {
-				utils.GetLogInstance().Warn("onViewChange failed to deserialize the multi signature for prepared payload", "error", err)
+				utils.GetLogger().Debug("m1 recvMsg payload read error", "error", err)
```
Comment (LeoHChen) [inline consensus/view_change.go]: Use Error.

File: consensus/view_change.go [hunk 4]
```diff
@@ -304,19 +296,12 @@ func (consensus *Consensus) onViewChange(msg *msg_pb.Message) {
...
-			bitmap := recvMsg.Payload[offset:]
-			//#### END Read payload data
-			aggSig := bls.Sign{}
-			_ = aggSig.Deserialize(multiSig)
-			mask, _ := bls_cosi.NewMask(consensus.PublicKeys, nil)
-			mask.SetMask(bitmap)
-			consensus.aggregatedPrepareSig = &aggSig
+			aggSig, mask, err := consensus.readSignatureBitmapPayload(recvMsg.Payload, 32)
+			if err != nil {
+				utils.GetLogger().Debug("readSignatureBitmapPayload fail", "error", err)
```
Comment (LeoHChen) [inline consensus/view_change.go]: Error.

File: consensus/view_change.go [hunk 5]
```diff
@@ -386,35 +371,19 @@ func (consensus *Consensus) onNewView(msg *msg_pb.Message) {
...
-		err := aggSig.Deserialize(multiSig)
-		if err != nil {
-			utils.GetLogInstance().Warn("onNewView unable to deserialize prepared message agg sig", "err", err)
-			return
-		}
-		mask, err := bls_cosi.NewMask(consensus.PublicKeys, nil)
+		aggSig, mask, err := consensus.readSignatureBitmapPayload(recvMsg.Payload, 32)
 		if err != nil {
-			utils.GetLogInstance().Warn("onNewView unable to setup mask for prepared message", "err", err)
+			utils.GetLogger().Debug("unable to read signature/bitmap", "error", err)
```
Comment (LeoHChen) [inline consensus/view_change.go]: Error

File: consensus/consensus_service.go [hunk 6]
```diff
@@ -603,3 +601,31 @@ func (consensus *Consensus) SetBlockNum(blockNum uint64) {
...
+
+// read the payload for signature and bitmap; offset is the beginning position of reading
+func (consensus *Consensus) readSignatureBitmapPayload(recvPayload []byte, offset int) (*bls.Sign, *bls_cosi.Mask, error) {
+	if offset+48 > len(recvPayload) {
+		return nil, nil, errors.New("payload not have enough length")
+	}
+	payload := append(recvPayload[:0:0], recvPayload...)
+	//#### Read payload data
+	// 48 byte of multi-sig
+	multiSig := payload[offset : offset+48]
```
Comment (LeoHChen) [inline consensus/consensus_service.go]: don't you need to check the length of payload?

File: consensus/consensus_service.go [hunk 7]
```diff
@@ -603,3 +601,31 @@ func (consensus *Consensus) SetBlockNum(blockNum uint64) {
...
+	offset += 48
+	// bitmap
+	bitmap := payload[offset:]
```
Comment (LeoHChen) [inline consensus/consensus_service.go]: check the length of payload, if it is less than offset, you will have panic.

Comment (chaosma) [reply consensus/consensus_service.go]: the bitmap payload is variable length, so I can only make sure the length is greater than 0

Comment (chaosma) [reply consensus/consensus_service.go]: checked in the beginning

File: consensus/consensus_service.go (hunk 6 above)
Comment (chaosma) [reply consensus/consensus_service.go]: see beginning of function

File: consensus/consensus_service.go (hunk 7 above)
Comment (rlan35) [reply consensus/consensus_service.go]: the first if check should be fine.

**Pull request merged**